lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,273 @@
+          gsub(%r(("Number\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]').
+          gsub(%r(("expirationDate\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]').
+          gsub(%r(("FirstName\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]').
+          gsub(%r(("LastName\\?"\s*:\s*\\?")[^"]*)i, '\1[FILTERED]')
+      end
+
+      private
+
+      def add_clerk(post, options)
+        post[:clerk] = {}
+        post[:clerk][:numericId] = options[:numeric_id]
+      end
+
+      def add_invoice(post, money, options)
+        post[:amount] = {}
+        post[:amount][:total] = money.to_f
+        post[:amount][:tax] = options[:tax].to_f || 0.0
+      end
+
+      def add_datetime(post, options)
+        post[:dateTime] = options[:date_time] || current_date_time
+      end
+
+      def add_transaction(post, options)
+        add_purchase_card(post, options)
+        add_override_bus_date(post, options)
+        add_amex(post, options)
+        add_auto(post, options)
+        add_card_on_file(post, options)
+      end
+
+      def add_card(post, credit_card, options)
+        post[:card] = {}
+        if credit_card.is_a?(CreditCard)
+          post[:card][:expirationDate] = "#{format(credit_card.month, :two_digits)}#{format(credit_card.year, :two_digits)}"
+          post[:card][:number] = credit_card.number
+          post[:card][:entryMode] = options[:entry_mode]
+          post[:card][:present] = options[:present]
+          post[:card][:securityCode] = {}
+          post[:card][:securityCode][:indicator] = 1
+          post[:card][:securityCode][:value] = credit_card.verification_value if credit_card.present?
+        else
+          post[:card] = {} if post[:card].nil?
+          post[:card][:token] = {}
+          post[:card][:token][:value] = credit_card
+        end
+      end
+
+      def add_customer(post, options)
+        if address = options[:billing_address]
+          post[:customer] = {}
+          post[:customer][:addressLine1] = address[:address1] if address[:address1]
+          name = address[:name].split(' ')
+          post[:customer][:firstName] = name[0]
+          post[:customer][:lastName] = name[1]
+          post[:customer][:postalCode] = address[:postal_code]
+        end
+      end
+
+      def add_purchase_card(post, options)
+        post[:transaction] = {}
+        post[:transaction][:invoice] = options[:invoice]
+        post[:transaction][:notes] = options[:notes] if options[:notes].present?
+      end
+
+      def add_override_bus_date(post, options)
+        post[:overrideBusDate] = options[:override_bus_date] if options[:override_bus_date].present?
+      end
+
+      def add_amex(post, options)
+        if options[:amex].present?
+          post[:amex] = {}
+          post[:amex][:propertyCode] = options[:amex][:property_code]
+        end
+      end
+
+      def add_auto(post, options)
+        if options[:auto].present?
+          post[:auto] = {}
+          post[:auto][:estimatedDays] = options[:auto][:estimated_days]
```

Comment:
Same as `property_code` above, pass it as `options[:estimated_days]` unless we have a good reason to do so.